dev/devicelab/lib/tasks/keyboard_hot_restart_test.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:path/path.dart' as path;
+
+import '../framework/devices.dart';
+import '../framework/framework.dart';
+import '../framework/task_result.dart';
+import '../framework/utils.dart';
+
+// Verifies that hot restart hides the keyboard if it is visible:
+//
+// 1. Launches an app that forces the keyboard to be visible
+// 2. Updates the app to no longer force the keyboard to be visible
+// 3. Hot restarts
+// 4. Checks that the keyboard is no longer visible
+TaskFunction createKeyboardHotRestartTest({
+  String? deviceIdOverride,
+  bool checkAppRunningOnLocalDevice = false,
+  List<String>? additionalOptions,
+}) {
+  final Directory appDir = dir(
+    path.join(flutterDirectory.path, 'dev/integration_tests/keyboard_hot_restart'),
+  );
+  final File mainFile = file(path.join(appDir.path, 'lib/main.dart'));
+
+  // This file is modified during the test and needs to be restored at the end.
+  final String oldContents = mainFile.readAsStringSync();
+  const String forceKeyboardOn = 'const bool forceKeyboard = true;';
+  const String forceKeyboardOff = 'const bool forceKeyboard = false;';
```

Comment:
is it just a raw string to be displayed? why does it look like a piece of code?

Reply:
I added a comment to clarify:

```dart
// When the test starts, the app forces the keyboard to be visible.
// The test turns off this behavior by mutating the app's source code from
// `forceKeyboardOn` to `forceKeyboardOff`.
// See: //dev/integration_tests/keyboard_hot_restart/lib/main.dart
```

Let me know if you have follow-up questions!

Reply:
wow changing the source code from the test, i've never seen this before. perhaps we can pass in an argument into the test instead.

Reply:
A test argument wouldn't work as we need to change the app's behavior at the middle of the test: first it should make the keyboard visible, but after the hot restart it shouldn't. Let me know if you'd like to do a call to talk through this further :)